src/FileFormats/NL/NL.jl

```diff
@@ -0,0 +1,788 @@
+            dest.x[x].in_nonlinear_constraint = true
+        end
+        for x in con.expr.nonlinear_terms
+            if x isa MOI.VariableIndex
+                dest.x[x].in_nonlinear_constraint = true
+            end
+        end
+    end
+    types = dest.types
+    for (x, v) in dest.x
+        if v.in_nonlinear_constraint && v.in_nonlinear_objective
+            push!(v.type == _CONTINUOUS ? types[1] : types[2], x)
+        elseif v.in_nonlinear_constraint
+            push!(v.type == _CONTINUOUS ? types[3] : types[4], x)
+        elseif v.in_nonlinear_objective
+            push!(v.type == _CONTINUOUS ? types[5] : types[6], x)
+        elseif v.type == _CONTINUOUS
+            push!(types[7], x)
+        elseif v.type == _BINARY
+            push!(types[8], x)
+        else
+            @assert v.type == _INTEGER
+            push!(types[9], x)
+        end
+    end
+    # However! Don't let Tables 3 and 4 fool you, because the ordering actually
+    # depends on whether the number of nonlinear variables in the objective only
+    # is _strictly_ greater than the number of nonlinear variables in the
+    # constraints only. Quoting:
+    #
+    #   For all versions, the first nlvc variables appear nonlinearly in at
+    #   least one constraint. If nlvo > nlvc, the first nlvc variables may or
+    #   may not appear nonlinearly in an objective, but the next nlvo – nlvc
+    #   variables do appear nonlinearly in at least one objective. Otherwise
+    #   all of the first nlvo variables appear nonlinearly in an objective.
+    #
+    # However, even this is slightly incorrect, because I think it should read
+    # "For all versions, the first nlvb variables appear nonlinearly." The "nlvo
+    # - nlvc" part is also clearly incorrect, and should probably read "nlvo -
+    # nlvb."
+    #
+    # It's a bit confusing, so here is the relevant code from Couenne:
+    # https://github.com/coin-or/Couenne/blob/683c5b305d78a009d59268a4bca01e0ad75ebf02/src/readnl/readnl.cpp#L76-L87
+    #
+    # They interpret this paragraph to mean the switch on nlvo > nlvc determines
+    # whether the next block of variables are the ones that appear in the
+    # objective only, or the constraints only.
+    #
+    # That makes sense as a design choice, because you can read them in two
+    # contiguous blocks.
+    #
+    # Essentially, what all this means is if !(nlvo > nlvc), then swap 3-4 for
+    # 5-6 in the variable order.
+    nlvc = length(types[3]) + length(types[4])
+    nlvo = length(types[5]) + length(types[6])
+    order_i = if nlvo > nlvc
+        [1, 2, 3, 4, 5, 6, 7, 8, 9]
+    else
+        [1, 2, 5, 6, 3, 4, 7, 8, 9]
+    end
+    # Now we can order the variables.
+    n = 0
+    for i in order_i
+        # Since variables come from a dictionary, there may be differences in
+        # the order depending on platform and Julia version. Sort by creation
+        # time for consistency.
+        for x in sort!(types[i]; by = y -> y.value)
```

Comment:
Not needed of we use `CleverDict` as it uses an `OrderedDict`

Reply:
No, this type ordering is very specific to NL files. It isn't the same as creation order.